examples/fishnet_api/main.py

```diff
@@ -1,414 +1,440 @@
+        for permission in await Permission.query(
+            timeseriesID=dataset.timeseriesIDs,
+            reader=execution.owner
+        )
+    }
+    requests = []
+    unavailable_timeseries = []
+    for ts in requested_timeseries:
+        if ts.owner == execution.owner:
+            continue
+        if not ts.available:
+            unavailable_timeseries.append(ts)
+        if requested_timeseries:
+            # continue to fetch all the unavailable timeseries for the denied response
+            continue
+        if ts.id_hash not in permissions:
+            # create permission request
+            requests.append(Permission.create(
+                timeseriesID=ts.id_hash,
+                algorithmID=execution.algorithmID,
+                owner=ts.owner,
+                reader=execution.owner,
+                status=PermissionStatus.REQUESTED,
+                executionCount=0,
+                maxExecutionCount=1,
+            ))
+        else:
+            # check if permission is valid
+            permission = permissions[ts.id_hash]
+            needs_update = False  # helper variable to avoid unnecessary updates
+            if permission.status == PermissionStatus.DENIED:
+                permission.status = PermissionStatus.REQUESTED
+                needs_update = True
+            if permission.maxExecutionCount <= permission.executionCount:
+                permission.maxExecutionCount = permission.executionCount + 1
+                permission.status = PermissionStatus.REQUESTED  # re-request permission
+                needs_update = True
+            if needs_update:
+                requests.append(permission.upsert())
+    if unavailable_timeseries:
+        execution.status = ExecutionStatus.DENIED
+        return RequestExecutionResponse(
+            execution=await Execution.create(**execution.dict()),
+            unavailableTimeseries=unavailable_timeseries
+        )
+    if requests:
+        new_permission_requests = await asyncio.gather(*requests)
+        execution.status = ExecutionStatus.REQUESTED
+        return RequestExecutionResponse(
+            execution=await Execution.create(**execution.dict()),
+            permissionRequests=new_permission_requests
+        )
+    else:
+        execution.status = ExecutionStatus.PENDING
+        return RequestExecutionResponse(
+            execution=await Execution.create(**execution.dict())
+        )
+
+
+@app.put("/permissions/approve")
+async def approve_permissions(permission_hashes: List[str]) -> FishnetResponseDataset:
+    """
+    Approve permission.
+    This EndPoint will approve a list of permissions by their item hashes
+    If an `id_hashes` is provided, it will change all the Permission status
+    to Granted.
+    """
+
+    ts_ids = []
+    requests = []
+    # TODO: Check signature to match with owner's
+
+    permission_record = await Permission.fetch(permission_hashes)
+    if permission_record:

```

Comment:
To reduce nesting, invert like so:
```suggestion
    if not permission_record:
        raise HTTPException(status_code=404, detail="No Permissions found with given hashes")
```